KAFKA-7131 Update release script to generate announcement email text

bibinss opened the pull request.

Commit:
KAFKA-7131 Update release script to generate announcement email text

bibinss commented:
@mjsax can you please take a look?

mjsax commented on release.py:
```diff
@@ -256,11 +260,105 @@ def command_stage_docs():
 
+def command_release_announcement_email():
+    release_version_num = raw_input('What is the current release version:')
+    previos_release_version_num = raw_input('What is the previous release version:')
```
If we run the script to do the actual release, we have this information already. It would be good to reuse this. Ie, we can keep this as-is, however add a second method that takes this information as parameters. This allow us to call the new method from here, after we collected the information, but also at the end of the regular execution of the script and pass in the information directly. Thus, if a committer does a release, it's not required to call the script again but the email template will be generated directly.

bibinss replied:
@mjsax couple of questions
1 - "If we run the script to do the actual release, we have this information already" - 
Do you mean to say that we can get the current and previos release versions from git tags? like first fetch all the tags, sort in descending order and pick first and second indexed values are current and previos versions?

2 - "This allow us to call the new method from here, after we collected the information, but also at the end of the regular execution of the script and pass in the information directly. Thus, if a committer does a release, it's not required to call the script again but the email template will be generated directly" - 
The script 'release.py' if run, by default generates the email content for RC release for voting. At this point we are not ready for sending release completion announcement. So, what is point of generating the announcement email content at the end of the regular execution of the script 'release.py'?
I mean, we need to send release announcement email only once voting is approved and we manually tag the release branch last commit with release version.

ewencp replied:
You definitely can determine this automatically from the existing tags. For anything with patch version > 0, it's trivial since you want the reference for previous version to be `patch_version - 1`. For the first release in a major.minor release line, you would need to figure out the correct previous major.minor.patch release and use that.

Normally I would say this is pretty easy, just list tags, find ones that match the right pattern, split segments on `.` characters, convert each to integers, and sort. However, this does get a bit messier with Kafka because of the switch in release numbering (from 4 digits in pre-1.0 to 3 digits in post-1.0), so you'd have to normalize to 4 digits, sort, then make sure you drop any extra digits from post-1.0 versions.

It'd be nice to get this all automated and the ergonomics of the script are nicer if they it is, but I wouldn't block merging this on that. This is still better than what committers do today, which is to just construct this all manually.

mjsax replied:
Should be ok to do either 3-digit or 4-digit code (for corresponding branches) ? No need to support both in one branch IMHO

mjsax reviewed (commented):
Sorry for late review and thanks for the PR. One initial comment. Did you test the script?

\cc @ewencp for second review

ewencp commented on release.py:
```diff
@@ -256,11 +260,105 @@ def command_stage_docs():
 
+def command_release_announcement_email():
+    release_version_num = raw_input('What is the current release version:')
+    previos_release_version_num = raw_input('What is the previous release version:')
```
nit: `previos` -> `previous`

bibinss replied:
this is done

ewencp commented on release.py:
```diff
@@ -256,11 +260,105 @@ def command_stage_docs():
     sys.exit(0)
 
+def command_release_announcement_email():
+    release_version_num = raw_input('What is the current release version:')
```
Should we do validation here like we do in the default staging command, i.e. like this https://github.com/apache/kafka/blob/trunk/release.py#L330-L338? Maybe we can refactor the validation/setting of global variable `release_version_parts` there so we can just reuse that bit of code?

bibinss replied:
yes, refactored the code

ewencp commented on release.py:
```diff
@@ -256,11 +260,105 @@ def command_stage_docs():
+    }
+
+    release_announcement_email = """
+        To: dev@kafka.apache.org, users@kafka.apache.org, kafka-clients@googlegroups.com
```
This should go to `announce@apache.org` as well, that's actually the most critical one as that makes it "official".

bibinss replied:
yes, done

ewencp commented on release.py:
```diff
@@ -256,11 +260,105 @@ def command_stage_docs():
+    print("*****************************************************************")
+    print()
+    print("Use the above template to send the announcement for the release to the mailing list.")
+    print("IMPORTANT: Note that there are still some substitutions that need to be made in the template:")
```
might want to also remind them that the email *must* come from their apache email address so it will be correctly delivered to `announce@apache.org` (see earlier comment). i.e. addressing this part of the wiki text that you are automating:

> You will need to use your apache email address to send out the email (otherwise, it won't be delivered to announce@apache.org)

bibinss replied:
this is done

bibinss replied:
yes, done

ewencp commented on release.py:
```diff
@@ -256,11 +260,105 @@ def command_stage_docs():
+        https://www.apache.org/dist/kafka/%(release_version)s/RELEASE_NOTES.html 
+ 
+
+        You can download the source and binary release (Scala 2.11 and Scala 2.12) from:        
```
Not sure how critical it is that we automate this, but these are configurable in the build now (see `availableScalaVersions`/`defaultScalaVersions` in `dependencies.gradle`). I'm a bit wary of trying to automate just because it requires parsing gradle code to do robustly, but at the same time, this seems like it'd be easy to have fall out of date.

I guess we could leave as is and hopefully people will be comprehensively `git grep`ing for places they need to change when adding/removing scala versions?

mjsax replied:
I double people will update this script correctly. We can only hope, that release managers verify this before sending the email...

As an alternative, we can also wildcard this, and let release manger insert those manually. Similar to `<DETAILS OF THE CHANGES>` above.

ewencp commented on release.py:
```diff
@@ -256,11 +260,105 @@ def command_stage_docs():
+    release_version_num = raw_input('What is the current release version:')
+    previos_release_version_num = raw_input('What is the previous release version:')
+    number_of_contributors = int(subprocess.check_output('git shortlog -sn --no-merges %s..%s | wc -l' % (previos_release_version_num, release_version_num) , shell=True))
+    contributors = subprocess.check_output("git shortlog -sn --no-merges %s..%s | cut -f2 | tr '\n' ',' | sed -e 's/,/, /g'" % (previos_release_version_num, release_version_num), shell=True)
```
This command has always left a trailing `,`. You could potentially omit the commands after the `cut` and just do a split/join in python that will give exactly what we want.

Also, not sure if it was intentional or not, but this command seems to elide the alphabetical sorting that's in the command on the wiki.

bibinss replied:
this is done

ewencp reviewed (commented):
@bibinss Some details, but this is mostly looking good to me. Thanks for taking this on! I'm a huge fan of anything we can do to further automate the (otherwise somewhat tedious) release process.

Commit:
incorporating review comments

bibinss commented:
@ewencp @mjsax I have added the logic to find the current and previous release from git tags. 
Also did work on most of your review comments. can you please take a look at the updated PR?

bibinss commented:
@ewencp @mjsax ^^

bibinss commented:
@ewencp  @mjsax can you please take a look at the updated PR?

mjsax commented on release.py:
```diff
@@ -256,11 +261,137 @@ def command_stage_docs():
+def validate_release_version_parts(version):
+    try:
+        version_parts = version.split('.')
+        if len(version_parts) != 3:
```
I like this.

We will be able to pack port this to `2.0`, `1.1`, and `1.0` branch -- not sure if there will ever be a `0.11.0.4` branch -- thus, might not be worth do back port further to support 4-digits.

ewencp replied:
we should plan on the possibility of another 0.11 release simply due to possible security fixes.

i agree validating release versions is nice. we can make this handle both old and new version schemes though. pre 1.0, there are 4 version parts but the first version part would always be `0`, so it's pretty easy to handle both cases.

mjsax replied:
For an 0.11 release, the committer can still put the email together manually... We need some cutoff point for back porting -- it's just a convenience improvement for doing a release, not a bug fix. Thus, 1.0.0 seems to be a clean cut-off point IMHO.

mjsax reviewed (commented):
Sorry for the delay. Was quite busy with the feature freeze deadline. Overall LGTM. Some more follow up.

Also: We need some additional cleanup in Line 679ff -- we should also call the new method there automatically.

Commit:
incorporating review comment: added placeholder for scala version

bibinss commented:
@mjsax 
- Added a place holder for scala verison in the generated email content and updated PR.
- On this comment " We need some additional cleanup in Line 679ff -- we should also call the new method there automatically." - Are you asking to call new method 'command_release_announcement_email'? why we need to do this at Line 679? at this line we are generating the 'Voting' email for RC right? why do we need to generate release complete email at this point?

bibinss commented:
@mjsax @ewencp can you please check

mjsax reviewed (approved):
LGTM.

Leave it to @ewencp to have a final look and merge.

ewencp commented on release.py:
```diff
@@ -256,11 +261,138 @@ def command_stage_docs():
+    if not user_ok("""Is the previous release %s ? (y/n):""" %previous_release_version_num):
+        previous_release_version_num = raw_input('What is the previous release version:')
+        validate_release_num(previous_release_version_num)
+    if release_version_num < previous_release_version_num :
```
This check isn't quite right since it would just check lexicographically (e.g. `'2.9.0' > '2.11.0'`, but the former is a smaller version than the latter). For now this will probably work fine since we don't hit minor version numbers that large, but eventually this won't work (e.g. major version numbers go > 9). (Similarly, the sort of tags isn't quite right either.)

ewencp committed:
Minor formatting cleanup

ewencp reviewed (approved):
LGTM, merging to trunk and back to 1.0

mjsax commented:
>  and back to 1.0

@ewencp Did you mean 2.0?

ewencp commented:
Well, I meant 1.0 when I wrote that, but then it didn’t cherrypick cleanly. Left note on the JIRA. In practice, shouldn’t matter since you can use release script from trunk to release older versions.
